kyma-project/third-party-images
# Grafana Image now based on Grafana Watcher

**Description**

Changes proposed in this pull request:

- Introduction of Grafana Watcher
   - Watches Grafana-Datasource directory in Grafana Pod
   - If file gets created, Grafana will be restarted to recognise newly added datasource; takes around 1-2 Seconds
   - Grafana endpoint has no downtime using this method
   - If user tries to access Grafana UI during this process reload, a loading page will be shown
   - There is still a bug on K8s side: Watch stream on configmaps/secrets shuts down after a certain time &#8594; Kiwiwgrid Sidecar checks every 60min if stream is healthy and recreates if necessary. This, is fine for now regarding @a-thaler 
- Grafana image now based on Grafana-Watcher

After this PR has been merged, the Kyma Kiwigrid K8s sidecar needs to run as a Container instead of an InitContainer. This is done with this [PR](https://github.com/kyma-project/kyma/pull/12486). Please also review this after reviewing this.

How to test:
1. Provision a cluster
2. Checkout this PR: https://github.com/kyma-project/kyma/pull/12486
3. Check if Grafana image is still set to the image of this PR, in the following file: `kyma/resources/monitoring/charts/grafana/values.yaml` Line 75: `PR-142`
4. Install Kyma using: `kyma deploy --source=local`; please use latest main
5. Create ConfigMap to create dashboard:

```yaml
apiVersion: v1
data:
  loki-datasource-2.yaml: |-
    apiVersion: 1
    datasources:
    - name: Loki-2
      type: loki
      access: proxy
      url: http://logging-loki.kyma-system:3100
      editable: true
      jsonData:
        maxLines: 1000
        derivedFields:
        - datasourceUid: Jaeger
          matcherRegex: "traceId=(\\w+)"
          name: TraceID
          url: "$${__value.raw}"
kind: ConfigMap
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"loki-datasource.yaml":"apiVersion: 1\ndatasources:\n- name: Loki\n  type: loki\n  access: proxy\n  url: http://logging-loki.kyma-system:3100\n  editable: true\n  jsonData:\n    maxLines: 1000\n    derivedFields:\n    - datasourceUid: Jaeger\n      matcherRegex: \"traceId=(\\\\w+)\"\n      name: TraceID\n      url: \"$${__value.raw}\""},"kind":"ConfigMap","metadata":{"annotations":{"reconciler.kyma-project.io/managed-by-reconciler-disclaimer":"DO NOT EDIT - This resource is managed by Kyma.\nAny modifications are discarded and the resource is reverted to the original state."},"labels":{"app":"loki","chart":"loki-2.5.2","grafana_datasource":"1","reconciler.kyma-project.io/managed-by":"reconciler","reconciler.kyma-project.io/origin-version":"local","release":"logging"},"name":"logging-loki-grafana-datasource","namespace":"kyma-system"}}
    reconciler.kyma-project.io/managed-by-reconciler-disclaimer: |-
      DO NOT EDIT - This resource is managed by Kyma.
      Any modifications are discarded and the resource is reverted to the original state.
  creationTimestamp: "2021-11-02T15:15:47Z"
  labels:
    app: loki
    chart: loki-2.5.2
    grafana_datasource: "1"
    reconciler.kyma-project.io/managed-by: reconciler
    reconciler.kyma-project.io/origin-version: local
    release: logging
  name: logging-loki-grafana-datasource-2
  namespace: kyma-system
  resourceVersion: "8691"
  selfLink: /api/v1/namespaces/kyma-system/configmaps/logging-loki-grafana-datasource
  uid: 62760082-06d5-4526-a73c-ecce8c2c4276
```
7. Portforward to grafana UI: `kpf svc/monitoring-grafana 8080:80 -n kyma-system`
8. Check in Grafana UI in Explore section, if Loki-2 is shown
9. Check Logs of `grafana-sc-datasources`: `k logs <monitoring grafana pod name> -c grafana-sc-datasources`
10. Should contain `Working on ADDED configmap kyma-system/logging-loki-grafana-datasource-2`; if not  please check at which time the configmaps have been updated, if the new is not showing up shorter than 1 hour after last update &#8594; ERROR

**Related issue(s)**

[Grafana datasource bug](https://github.com/kyma-project/kyma/issues/12403)
